### PR TITLE
Adaptation to the Peng et Al framework

### PR DESCRIPTION
**Pull request by matol-16**

### CODE REVIEW

File: byzfl/attacks/attacks.py
Comment (matol-16) [inline byzfl/attacks/attacks.py:1]: Il me semble que le static label flipping le correspond pas exactement à celui du papier *mean is more robust*. Dans mean is more robust, le flip est déterministe:  targets = 9 - targets. ON pourrait peut être ajouter une variable "random" pour effectuer un flip aléatoire dans un deuxième temps.

File: byzfl/benchmark/evaluate_results.py
Comment (matol-16) [inline byzfl/benchmark/evaluate_results.py:1]: Ici il me semble que dans chacune des méthodes que tu implémentes, tu plottes encore le accuracy path  (via tab_acc)? Par ailleurs, par exemple quand tu plots  la feature variance, pas besoin de le faire pour chaque attaque ou aggrégateur puisque a priori la feature variance dépend uniquement de la distribution. ON peut en discuter sur whatsapp. Concernant le rho (pas encore implémenté de ce que je comprends), je pense qu'on peut reprendre directement les formules du papier.

File: byzfl/benchmark/train.py [hunk 1]
```diff
@@ -166,22 +166,29 @@ def start_training(params):
...
+        PoisoningClient({"model_name": params_manager.get_model_name(),
+            "device": params_manager.get_device(),
+            "optimizer_name": params_manager.get_optimizer_name(),
+            "learning_rate": params_manager.get_learning_rate(),
+            "loss_name": params_manager.get_loss_name(),
+            "weight_decay": params_manager.get_honest_clients_weight_decay(),
+            "milestones": params_manager.get_milestones(),
+            "learning_rate_decay": params_manager.get_learning_rate_decay(),
+            "LabelFlipping": "LabelFlipping" == params_manager.get_attack_name(),
+            "training_dataloader": client_dataloaders[i],
```
Comment (matol-16) [inline byzfl/benchmark/train.py]: ici, il me semble que le poisoned cleints récupère les mêmes dataloaders que les honest clients. Il faudrait probablement créer nb_honest+nb_byz_clients dataloaders, et assigner les nb_byz_clients derniers dataloaders aux clients empoisonnés

File: byzfl/benchmark/train.py [hunk 2]
```diff
@@ -209,19 +217,19 @@ def start_training(params):
     
     training_algorithm_name = params_manager.get_training_algorithm_name()
 
-    if training_algorithm_name not in ["DSGD", "FedAvg"]:
+    if training_algorithm_name not in ["DSGD"]:
         raise ValueError(f"Training algorithm {training_algorithm_name} not supported, supported algorithms are 'DSGD' and 'FedAvg'")
```
Comment (matol-16) [inline byzfl/benchmark/train.py:242]: message d'erreur à changer probabelemnt

File: byzfl/benchmark/train.py [hunk 3]
```diff
@@ -263,81 +271,71 @@ def start_training(params):
...
 
-            # Deal with Label Flipping Attack
-            attack_input = (
-                [client.get_flat_flipped_gradients() for client in honest_clients]
-                if label_flipping_attack
-                else honest_gradients
-            )
+            # Apply poisonning attack
+            for poisonned_client in poisonned_clients:
+                poisonned_client.compute_gradients()
```
Comment (matol-16) [inline byzfl/benchmark/train.py]: ici, ca veut dire qu'on n'enregistre pas les variances et la loss des byzantins ?

File: byzfl/fed_framework/models.py
Comment (matol-16) [inline byzfl/fed_framework/models.py:1]: ici, il faudrait peut être changer les noms des modèles ajoutés (type: CNN_MNIST_PENG) ou un truc comme ca, pcq il y a des modèles implémentés par byzfl avec le même nom mais en miiuscule (genre cnn_mnist)

File: byzfl/benchmark/train.py
Comment (matol-16) [inline byzfl/benchmark/train.py:1]: est-ce que tu penses qu'il serait utile d'ajouter un boolean pour gérer l'enregistrement des variances etc ? Je me dis que quand on passera sur des modèles plus gros, le faire à chaque training step sera coûteux

File: byzfl/benchmark/train.py (hunk 1 above)
Comment (matol-16) [reply byzfl/benchmark/train.py]: Ca me fait penser: est-ce que on a pris en compte dans le reste de byzfl qu'on considère 2 types de clients désormais, et pas uniquement les honest clients soumis à des attaques omniscientes ? J'imagine qu'il y a des spillovers notamment pour les plots: il faudra pouvoir différencier les points des byzantins et des honnêtes au moment de plotter.

File: byzfl/benchmark/train.py (hunk 3 above)
Comment (Ayoub-Gourari) [reply byzfl/benchmark/train.py]: On a pas besoin de la loss des byzantins (même pas de celle des honnêtes en vrai). Pour les variances je viens de régler.

File: byzfl/benchmark/train.py (hunk 2 above)
Comment (Ayoub-Gourari) [reply byzfl/benchmark/train.py:242]: Pas forcément je crois. Y'a moyen qu'on revienne à "FedAvg" pour LFighter (cf. commentaire sur LFighter sur le markdown des tâches)

File: byzfl/benchmark/train.py (hunk 1 above)
Comment (Ayoub-Gourari) [reply byzfl/benchmark/train.py]: Il me semble que ce n'est pas grave, la seule différence c'est dans le calcul des "scatterings" qui est gérée dans la ligne 413 -> 419 du fichier train. (On a pas besoin de prendre en compte la norme des moyennes des features, ni de leur accuracies dans nos expériences)

File: byzfl/benchmark/evaluate_results.py
Comment (Ayoub-Gourari) [reply byzfl/benchmark/evaluate_results.py:1]: Oui c'est ça. Par rapport au plot de la feature variance, on peut appeler la fonction une fois par distribution (il me semble que nous qui choisiront quand on va les appeler dans l'autre repo). Concernant les convergence rates, j'attends qu'on avance et qu'on se mette d'accords sur la partie théorique pour pouvoir les implémenter, normalement les papiers introduisants chaque aggregateur robuste évaluent son rho.

File: byzfl/attacks/attacks.py
Comment (Ayoub-Gourari) [reply byzfl/attacks/attacks.py:1]: Fait !

File: byzfl/benchmark/evaluate_results.py
Comment (matol-16) [reply byzfl/benchmark/evaluate_results.py:1]: Mais donc je ne comprenais exactement pourquoi tu replottais à chaque fois le accuracy path, même dans la fonction de plot de la la feature mean: plot_maximum_regular_feature_mean

File: byzfl/benchmark/train.py (hunk 1 above)
Comment (matol-16) [reply byzfl/benchmark/train.py]: Ca marche ! en revanche il me semble qu'on a quand même le soucis des dataloaders partagés non ? (cf mon 1er commentaire). Genre le datadistributor va générer nb_honest dataloaders. Il faudrait probablement changer cela à nb_honest+nb_poisonned pour avoir bien une distribution différente pour chaque client.

Comment (Ayoub-Gourari) [reply byzfl/benchmark/train.py]: Oui pardon je l'ai résolu. La class de data loader prends en arguments mtn le nombre total de workers

File: byzfl/attacks/attacks.py [hunk 4]
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
...
+    """
 
-    def __init__(self):
-        pass        
+    def __init__(self, p):
+        self.p = p        
+    
+    def __call__(self, model, inputs, outputs):      
+        classes = list(set(outputs))
+        mapping = {outputs : model.nb_classes - 1 - outputs for outputs in classes}
```
Comment (slivering) [inline byzfl/attacks/attacks.py]: Ok, mais on pourrait aussi rendre la permutation configurable dans `__init__` (spécifier `mapping` en argument de l'attaque)

File: byzfl/attacks/attacks.py [hunk 5]
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
...
+
+        flipped_outputs = [mapping[output] for output in outputs[:int(self.p*len(outputs))]] + outputs[int(self.p*len(outputs)):]
```
Comment (slivering) [inline byzfl/attacks/attacks.py]: Ok, mais on pourrait aussi effectuer le label flipping sur un sous-ensemble **aléatoire** des indices avec `np.random.Generator.choice(len(outputs), int(p * len(outputs)))`

(`np.random.Generator` : à configurer avec une seed)

File: byzfl/benchmark/managers.py [hunk 6]
```diff
@@ -111,18 +111,18 @@ def save_honest_scattering(self, honest_scattering_list, training_seed, data_dis
         file_path = os.path.join(acc_dir, f"honest_gradients_scattering.txt")
         np.savetxt(file_path, honest_scattering_list, fmt="%.4f", delimiter=",")
 
-    def save_byzantine_scattering(self, byzantine_scattering_list, training_seed, data_dist_seed):
+    def save_poisonned_scattering(self, poisonned_scattering_list, training_seed, data_dist_seed):
```
Comment (slivering) [inline byzfl/benchmark/managers.py]: Un seul n à poisoned (à appliquer à chaque instance du mot)

File: byzfl/attacks/attacks.py [hunk 7]
```diff
@@ -1121,11 +1122,33 @@ def __call__(self, honest_vectors):
...
+
+class DynamicLabelFlipping(object):
+    """
+    Implementation of dynamic label flipping from Peng & al.
+    """
+
+    def __init__(self, p):
+        self.p = p
+
+    def __call__(self, model, inputs, outputs):
```
Comment (slivering) [inline byzfl/attacks/attacks.py]: Cette méthode n'utilise pas le paramètre `self.p`...

File: byzfl/fed_framework/poisoning_client.py [hunk 8]
```diff
@@ -189,20 +194,17 @@ def apply_attack(self, inputs, targets):
             raise ValueError("Inputs and targets lengths mismatch")
 
         inputs, targets = deepcopy((inputs, targets))
-        if self.no_attack:
+        if self.no_attack or self.attack == None or self.p == 0.0:
             return inputs, targets
-
-        n = len(inputs)
-        m = int(self.p * n)
-        poisoned_indices = np.random.choice(n, m, replace=False)
```
Comment (slivering) [inline byzfl/fed_framework/poisoning_client.py:197]: Pourquoi avoir déplacé ce code dans chaque attaque par label flipping ?
L'intérêt de ces lignes était d'appliquer l'attaque par empoisonnement sur tous les `poisoned_indices`. Cela évitait de dupliquer cette logique.

File: byzfl/aggregators/aggregators.py [hunk 9]
```diff
@@ -675,46 +676,101 @@ def clusters_dissimilarity(self, clusters):
         ds1 = n1/m * (1 - np.mean(mincs1))
         return ds0, ds1
     
-    def __call__(self, vectors):
-        m = len(vectors)
-        dw = [[] for _ in range(m)]
+    # Get average weights
```
Comment (slivering) [inline byzfl/aggregators/aggregators.py]: `@staticmethod` pour appeler `LFighter.average_weights(w, marks)`.

File: byzfl/aggregators/aggregators.py [hunk 10]
```diff
@@ -675,46 +676,101 @@ def clusters_dissimilarity(self, clusters):
...
+        for key in w_avg.keys():
+            w_avg[key] = w_avg[key] * marks[0]
+        for key in w_avg.keys():
+            for i in range(1, len(w)):
+                w_avg[key] += w[i][key] * marks[i]
+            w_avg[key] = w_avg[key] *(1/sum(marks))
+        return w_avg
+
+    # Adapt to gradients ...
+    def aggregate(self, global_model, local_models, ptypes):
```
Comment (slivering) [inline byzfl/aggregators/aggregators.py]: Non applicable car l'on ne dispose pas de state dicts mais de gradients 1D. Il faudra implémenter `__call__` en adaptant l'implémentation originale.

File: byzfl/aggregators/aggregators.py [hunk 11]
```diff
@@ -691,7 +691,7 @@ def average_weights(w, marks):
         return w_avg
 
     # Adapt to gradients ...
-    def aggregate(self, global_model, local_models, ptypes):
+    def __call__(self, global_model, local_models):
```
Comment (matol-16) [inline byzfl/aggregators/aggregators.py]: Ici, tu calls avec global_model + local_models en paramètre. Ca ne fonctionne pas avec train.py, qui appelle uniquement un aggregateur avec en argument un objet "vector" (les params du modèles, ou les gradients).

Il faut donc probabelment faire une exception pour Lfighter, pour pouvoir lui feed les modèles locaux et globaux... On peut peut être s'inspirer de ce que font Peng et al

Comment (Ayoub-Gourari) [reply byzfl/aggregators/aggregators.py]: J'ai push une modification dans une nouvelle branche Lfighter_adaptation (désolé, on travaillais au même temps et du coup je suis devenu asynchro).

File: byzfl/fed_framework/client.py [hunk 12]
```diff
@@ -123,14 +154,55 @@ def _backward_pass(self, inputs, targets, train_acc=False):
...
+
+            # Compute the gradients for each data point
+            individual_gradients = []
+            for input, target in zip(inputs, targets):
+                self.model.zero_grad()
+                output = self.model(input.unsqueeze(0))  # Add batch dimension
+                loss = self.criterion(output, target.unsqueeze(0))  # Add batch dimension
+                loss.backward()
+                
+                grad_vector = torch.cat([p.grad.view(-1) for p in self.model.parameters() if p.grad is not None])
```
Comment (slivering) [inline byzfl/fed_framework/client.py:176]: `grad_vector = self.get_flat_gradients()` devrait être équivalent

**Review by slivering — COMMENTED** [review]
Quelques remarques principales :

1. Laisser la responsabilité à `PoisoningClient` de sélectionner une fraction **aléatoire** des données à empoisonner (rendre la seed configurable)
2. Calcul douteux de `dw` et `db` dans LFighter
3. `Server.update_model_with_models` suspicieux (problème de signature avec les agrégateurs qui prennent des gradients en argument)
